arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.cpp

```diff
@@ -0,0 +1,714 @@
+    futures::Future<std::vector<Step*>> futureEnds = _provider.fetch(looseEnds);
+    futureEnds.get();
+    // Notes for the future:
+    // Vertices are now fetched. Thnink about other less-blocking and batch-wise
+    // fetching (e.g. re-fetch at some later point).
+    // TODO: Discuss how to optimize here. Currently we'll mark looseEnds in
+    // fetch as fetched. This works, but we might create a batch limit here in
+    // the future. Also discuss: Do we want (re-)fetch logic here?
+    // TODO: maybe we can combine this with prefetching of paths
+  }
+}
+
+template<class QueueType, class PathStoreType, class ProviderType,
+         class PathValidator>
+auto WeightedTwoSidedEnumerator<QueueType, PathStoreType, ProviderType,
+                                PathValidator>::Ball::hasBeenVisited(Step const&
+                                                                         step)
+    -> bool {
+  if (_visitedNodes.contains(step.getVertex().getID())) {
+    return true;
+  }
+  return false;
+}
+
+template<class QueueType, class PathStoreType, class ProviderType,
+         class PathValidator>
+auto WeightedTwoSidedEnumerator<QueueType, PathStoreType, ProviderType,
+                                PathValidator>::Ball::
+    computeNeighbourhoodOfNextVertex(Ball& other, CandidatesMap& candidates,
+                                     double& matchPathLength) -> void {
+  // Pull next element from Queue
+  // Do 1 step search
+  TRI_ASSERT(!_queue.isEmpty());
+  if (!_queue.hasProcessableElement()) {
+    std::vector<Step*> looseEnds = _queue.getLooseEnds();
+    futures::Future<std::vector<Step*>> futureEnds = _provider.fetch(looseEnds);
+
+    // Will throw all network errors here
+    auto&& preparedEnds = futureEnds.get();
+
+    TRI_ASSERT(preparedEnds.size() != 0);
+    TRI_ASSERT(_queue.hasProcessableElement());
+  }
+
+  auto step = _queue.pop();
+  if (hasBeenVisited(step)) {
+    return;
+  }
+  auto previous = _interior.append(step);
+
+  auto verifyStep = [&](Step currentStep) {
+    TRI_ASSERT(!hasBeenVisited(currentStep));
+    ValidationResult res = _validator.validatePath(currentStep);
+
+    if (!res.isPruned()) {
+      // Add the step to our shell
+      _queue.append(std::move(currentStep));
+    }
+  };
+
+  _visitedNodes.emplace(step.getVertex().getID(), previous);
+  if (other.hasBeenVisited(step)) {
+    // Shortest Path Match
+    matchPathLength = other.matchResultsInShell(step, candidates, _validator);
+  }
+
+  _provider.expand(step, previous, [&](Step n) -> void {
+    if (!hasBeenVisited(n)) {
+      verifyStep(std::move(n));
+    }
+  });
+}
+
+template<class QueueType, class PathStoreType, class ProviderType,
+         class PathValidator>
+void WeightedTwoSidedEnumerator<
+    QueueType, PathStoreType, ProviderType,
+    PathValidator>::Ball::testDepthZero(Ball& other,
+                                        CandidatesMap& candidates) {
+  // TODO: Check - is this even required?
```

Comment:
This must be checked before we merge back. 
The min and max values are being set here: 

ShortestPathNode:413
```
  arangodb::graph::TwoSidedEnumeratorOptions enumeratorOptions{
      0, std::numeric_limits<size_t>::max(),
      arangodb::graph::PathType::Type::ShortestPath};
```